### PR TITLE
Add missing indexes from #7234

### PR DESCRIPTION
**Pull request by SuperTux88**
I added the indexes reviewed by @CSammy from https://github.com/diaspora/diaspora/pull/7234#issuecomment-267187943

I also refactored the `AccountDeleter` to work with the person, that can be used in #6750 too.

I didn't add the unique index for the `guid` in the `participations` table. We already have an index there and we have a unique index for the `target` which should be enough. I planned to remove the `guid` column anyway soon, because we don't really need it, so there is no need to change the existing index to a unique index there.

### CODE REVIEW

File: lib/account_deleter.rb [hunk 1]
```diff
@@ -69,10 +68,6 @@ def delete_standard_person_associations
     end
   end
 
-  def disconnect_contacts
```
Comment (cmrd-senya) [inline lib/account_deleter.rb]: Why did you remove this?

**Comment by cmrd-senya** [comment]
Your refactorings conflict with the changes from #6750. What do you think we should merge first?

File: lib/account_deleter.rb (hunk 1 above)
Comment (SuperTux88) [inline lib/account_deleter.rb]: See commit-message: Also remove `disconnect_contacts` methods, because contacts are already removed with aspects memberships in `before_destroy`.

I had a problem after changing this to use `person` instead of `diaspora_handle`, because it tried to do this twice, but the second time it failed, because `contact.user` was nil.

**Comment by SuperTux88** [comment]
> Your refactorings conflict with the changes from #6750. What do you think we should merge first?

I know, that's why I already mentioned it in the PR description. Since you changed `AccountDeleter` in #6750 to handle persons too, you can use the change here, and you don't need [this if](https://github.com/cmrd-senya/diaspora/blob/5dbc1b0adf9f1f593024173dd59de9b4da1708fa/lib/account_deleter.rb#L21-L25) anymore there. So I think it's cleaner to merge this first, and the conflicts should be easy to resolve (I only removed `diaspora_handle`, and since you don't rely on that, that shouldn't be a problem for your PR).

File: lib/account_deleter.rb (hunk 1 above)
Comment (SuperTux88) [inline lib/account_deleter.rb]: And the tests ensures that `contacts` are empty after the `AccountDeleter` run, so that shouldn't be a problem when something changes in the future.

**Comment by SuperTux88** [comment]
Oh, I broke the tests last minute, because pronto told me so :( I'll fix that ...

**Comment by cmrd-senya** [comment]
Usage of `person` in `AccountDeleter` is what I originally was trying to do in #6783 (which I later included to #6726).

@denschub disapproved this change back then: https://github.com/diaspora/diaspora/pull/6783#r60526640

**Comment by SuperTux88** [comment]
Hmm, I don't see why we shouldn't use `person` here, we already have it and can continue using it. It doesn't make sense to query an entity again.

**Comment by cmrd-senya** [comment]
> I know, that's why I already mentioned it in the PR description. Since you changed AccountDeleter in #6750 to handle persons too, you can use the change here, and you don't need this if anymore there. So I think it's cleaner to merge this first, and the conflicts should be easy to resolve (I only removed diaspora_handle, and since you don't rely on that, that shouldn't be a problem for your PR).

Okay, I need to review this PR then to check it doesn't affect anything that I use in migration.

#6750 also changes some associations in objects and in `account_deleter` (method `normal_ar_person_associates_to_delete` and friends) so that's not only `diaspora_handle`. I should check this I think.

**Comment by cmrd-senya** [comment]
> Hmm, I don't see why we shouldn't use person here, we already have it and can continue using it. It doesn't make sense to query an entity again.

That's what I thought either, but I changed it to this `if` approach just because of this comment.

**Comment by denschub** [comment]
> Hmm, I don't see why we shouldn't use person here, we already have it and can continue using it.

If we already have inside the current worker, that's totally fine, but passing around objects between worker causes them to get serialized to redis, which is a bit of a bummer. However, it does not seem like we end up with that here, so I may made a brainfart in the comment back then.

**Comment by SuperTux88** [comment]
> Okay, I need to review this PR then to check it doesn't affect anything that I use in migration.

I checked while editing this (but that was before your latest push), and you extacted the `close_user` to a new method, but that is an easy change (you need to remove `disconnect_contacts` there too) and can be resolved.

I didn't change anything `normal_ar_person_associates_to_delete`, but I changed from `%i()` to `%i[]`, because it was really hard to read because rubymine adds curly yellow lines when there is a rubocop error, and I don't see if there is a `_` or a ` ` ;) I but I can revert that back if you like.

**Comment by SuperTux88** [comment]
Also, [tests fixed](https://jenkins.coding4.coffee/job/diaspora-2.4-other/884/) (travis is still busy with your PR) :)

**Review by skyschub — APPROVED** [review]
Take this as approved once travis is green.

File: spec/integration/account_deletion_spec.rb [hunk 2]
```diff
@@ -1,19 +1,21 @@
 describe "deleteing account", type: :request do
   def account_removal_method
-    AccountDeleter.new(subject.diaspora_handle).perform!
+    AccountDeleter.new(subject).perform!
     subject.reload
   end
 
   context "of local user" do
-    subject(:user) { FactoryGirl.create(:user_with_aspect) }
```
Comment (cmrd-senya) [inline spec/integration/account_deletion_spec.rb]: I'd rahter not change this part here since it will be hard to rebase. Maybe it's better to add this change on top of my PR instead?

Comment (cmrd-senya) [inline spec/integration/account_deletion_spec.rb]: (this applies to most of the changes in this file, maybe excluding line 3)

**Review by cmrd-senya — APPROVED** [review]

File: lib/account_deleter.rb (hunk 1 above)
Comment (cmrd-senya) [inline lib/account_deleter.rb]: Actually `before_destroy` block in `AspectMembership` doesn't delete contacts if it's `sharing: true`. So it's not the same thing as `disconnect_contacts` in `AccountDeleter`. I have no idea why it worked for you, but it now fails with contacts not empty for me.

Here is my pry session output that shows what happens to a contact actually:
```
From: /home/senya/source/diaspora/app/models/aspect_membership.rb @ line 14 :

     9:   has_one :user, :through => :contact
    10:   has_one :person, :through => :contact
    11: 
    12:   before_destroy do
    13:     binding.pry
 => 14:     if self.contact && self.contact.aspects.size == 1
    15:       self.user.disconnect(self.contact)
    16:     end
    17:     true
    18:   end
    19: 

[2] pry(#<AspectMembership>)> self.contact.aspects.size
=> 1
[4] pry(#<AspectMembership>)> user.contacts
=> [#<Contact:0x0000000dfeff08
  id: 1105,
  user_id: 1139,
  person_id: 4295,
  created_at: Mon, 14 Aug 2017 01:56:40 UTC +00:00,
  updated_at: Mon, 14 Aug 2017 01:56:40 UTC +00:00,
  sharing: true,
  receiving: true>]
[5] pry(#<AspectMembership>)> next

From: /home/senya/source/diaspora/app/models/aspect_membership.rb @ line 15 :

    10:   has_one :person, :through => :contact
    11: 
    12:   before_destroy do
    13:     binding.pry
    14:     if self.contact && self.contact.aspects.size == 1
 => 15:       self.user.disconnect(self.contact)
    16:     end
    17:     true
    18:   end
    19: 
    20:   def as_json(opts={})

[5] pry(#<AspectMembership>)> next

From: /home/senya/source/diaspora/app/models/aspect_membership.rb @ line 17 :

    12:   before_destroy do
    13:     binding.pry
    14:     if self.contact && self.contact.aspects.size == 1
    15:       self.user.disconnect(self.contact)
    16:     end
 => 17:     true
    18:   end
    19: 
    20:   def as_json(opts={})
    21:     {
    22:       :id => self.id,

[6] pry(#<AspectMembership>)> user.contacts.reload
=> [#<Contact:0x0000000f50cfa8
  id: 1105,
  user_id: 1139,
  person_id: 4295,
  created_at: Mon, 14 Aug 2017 01:56:40 UTC +00:00,
  updated_at: Mon, 14 Aug 2017 01:57:25 UTC +00:00,
  sharing: true,
  receiving: false>]
```

Comment (cmrd-senya) [inline lib/account_deleter.rb]: Don't you mind me reverting the change with `disconnect_contacts` in the account deleter?

Comment (SuperTux88) [inline lib/account_deleter.rb]: You are right ... when the other person still shares with you, the contact doesn't get deleted ... So the data generator probably only creates one-way contacts? because user.contacts were empty after reload.

Comment (SuperTux88) [inline lib/account_deleter.rb]: No, one half of the contacts are deleted via `delete_contacts_of_me` and the second half is deleted with deleting aspects in `delete_standard_user_associations`. And when I wanted to delete it a third time in `disconnect_contacts` it fails with:
```
NoMethodError: undefined method `disconnect' for nil:NilClass
./app/models/aspect_membership.rb:14:in `block in <class:AspectMembership>'
...
./lib/account_deleter.rb:73:in `disconnect_contacts'
./lib/account_deleter.rb:37:in `block in perform!'
...
```

When I do `user.contacts.reload.destroy_all` in `disconnect_contacts` it works, but there are no contacts left.

Comment (cmrd-senya) [inline lib/account_deleter.rb]: Ah, I see, I missed that before, but I made a little change in `DataGenerator` for #6750 which actually created a double-sided contact: https://github.com/diaspora/diaspora/pull/6750/files#diff-4af80ea1ce63ed0a143b1b026bdb7eec

When I made methods for `DataGenerator` that are used in the account deletion spec I followed the spec that was there before. This spec didn't create any double-sided contacts. There are methods in `DataGenerator` which can make a double-sided contact but I didn't call them for the account deletion spec, only for exporter. But with the change in #6750 it is now different.

Comment (cmrd-senya) [inline lib/account_deleter.rb]: > No, one half of the contacts are deleted via delete_contacts_of_me and the second half is deleted with deleting aspects in delete_standard_user_associations

No, `delete_contacts_of_me` are contacts of local people who have us in their aspects. The thing that isn't deleted are contacts where `user` is us, and `person` is someone else, while `delete_contacts_of_me` deletes contacts where we're a person. So that's a different thing.

Comment (SuperTux88) [inline lib/account_deleter.rb]: You are right, when I add your change from the DataGenerator, it fails without `disconnect_contacts` ... We probably need better tests, because without the additional double-sided contact, it fails without `.reload`. I created #7538 to fix that, that should now cover both problems. 

I need to sleep now, because I'm tired and I need to work tomorrow, but I'll have a closer look at this tomorrow, to improve the tests. Or if you have the time to improve the tests in your PR, you can also do it.

@denschub: can you merge #7538 as soon as travis is green?

Comment (cmrd-senya) [inline lib/account_deleter.rb]: > We probably need better tests

What tests exactly? Some more tests for this method in `spec/lib/account_deleter_spec.rb`?

Comment (cmrd-senya) [inline lib/account_deleter.rb]: @SuperTux88, good night and thank you for your work!

Comment (SuperTux88) [inline lib/account_deleter.rb]: In this case: tests with different combinations of contacts. Without `disconnect_contacts` it fails when you have double sided contacts, for me it failed without `.reload`. So something that covers that.

But I just tested, when I remove the `.reload` it now also fails with the updated DataGenerator, so it's maybe already enough. I just want to make sure that we don't remove the `.reload`  or `disconnect_contacts` and something is broken again without a failing test.

But I'll have a closer look at this again tomorrow. #7538 should fix what I broke and as it looks it also makes the tests failing if I break something again, so for now that's good enough ;) 😴
